unit_tests/UnitTestLinearSystem.h

```diff
@@ -10,16 +10,98 @@
 
 #include "LinearSystem.h"
 #include "EquationSystem.h"
+#include "utils/CreateDeviceExpression.h"
 
 namespace unit_test_utils {
 
-class TestLinearSystem : public sierra::nalu::LinearSystem
+template<typename InputLhsType, typename InputRhsType,
+         typename LHSView, typename RHSView>
+KOKKOS_FUNCTION
+void assign(const InputLhsType& inputLhs, const InputRhsType& inputRhs,
+            LHSView lhs, RHSView rhs)
+{
+    for(size_t i=0; i<inputRhs.extent(0); ++i) {
+      rhs(i) = inputRhs(i);
+    }
+    for(size_t i=0; i<inputLhs.extent(0); ++i) {
+      for(size_t j=0; j<inputLhs.extent(1); ++j) {
+        lhs(i,j) = inputLhs(i,j);
+      }
+    }
+}
+
+using LHSView = Kokkos::View<double**,Kokkos::LayoutRight, sierra::nalu::MemSpace>;
+using RHSView = Kokkos::View<double*,Kokkos::LayoutRight, sierra::nalu::MemSpace>;
+
+class TestCoeffApplier : public sierra::nalu::CoeffApplier
 {
 public:
+  KOKKOS_FUNCTION
+  TestCoeffApplier(LHSView& lhs, RHSView& rhs, unsigned numContributionsToAccept = 1)
+  : lhs_(lhs), rhs_(rhs), devicePointer_(nullptr),
+    numContributionsToAccept_(numContributionsToAccept), numContributions_(0)
+  {}
 
- TestLinearSystem( sierra::nalu::Realm &realm, const unsigned numDof, sierra::nalu::EquationSystem *eqSys)
+  KOKKOS_FUNCTION
+  TestCoeffApplier(const TestCoeffApplier&) = default;
+
+  KOKKOS_FUNCTION
+  ~TestCoeffApplier()
+  {
+    if (this != devicePointer_) {
```

Comment:
In `NGPKernel` I moved this out of the destructor to avoid warnings regarding `kokkos_free_on_device`, how did you avoid those warnings here? I would recommend providing a `free_on_device` method that can be called from `TestLinearSystem`.

Reply:
Yes I think your approach is better, I'll clean that up. I do have warnings about that.